openshift/default-catalog-consistency/pkg/extract/extract.go

```diff
@@ -0,0 +1,223 @@
+package extract
+
+// TODO: Create a utility lib with the code to pull and extract the images
+// used in this file to be reused for this project and others.
+// More info: https://issues.redhat.com/browse/OPRUN-3919
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	imgcopy "github.com/containers/image/v5/copy"
+	"github.com/containers/image/v5/docker"
+	"github.com/containers/image/v5/docker/reference"
+	"github.com/containers/image/v5/manifest"
+	"github.com/containers/image/v5/oci/layout"
+	"github.com/containers/image/v5/pkg/compression"
+	"github.com/containers/image/v5/signature"
+	"github.com/containers/image/v5/types"
+	"github.com/containers/storage/pkg/archive"
+	"oras.land/oras-go/v2"
+	"oras.land/oras-go/v2/content/oci"
+)
+
+// ExtractedImage represents the extracted OCI content and its temporary directory.
+type ExtractedImage struct {
+	Store  oras.ReadOnlyTarget
+	TmpDir string
+	Tag    string
+}
+
+// Cleanup function to remove the temporary directory
+func (r *ExtractedImage) Cleanup() {
+	if r.TmpDir != "" {
+		if err := os.RemoveAll(r.TmpDir); err != nil {
+			fmt.Printf("failed to remove temp dir: %v\n", err)
+		}
+	}
+}
+
+// UnpackImage pulls the image, extracts it to disk, and opens it as an OCI store.
+func UnpackImage(ctx context.Context, imageRef, name string) (res *ExtractedImage, err error) {
+	tmpDir, err := os.MkdirTemp("", fmt.Sprintf("oci-%s-", name))
+	if err != nil {
+		return nil, fmt.Errorf("create temp dir: %w", err)
+	}
+
+	var digestTag string
+
+	res, err = func() (*ExtractedImage, error) {
+		srcRef, err := docker.ParseReference("//" + imageRef)
+		if err != nil {
+			return nil, fmt.Errorf("parse image ref: %w", err)
+		}
+
+		// Force image resolution to Linux platform to avoid OS mismatch errors
+		// during testing on Mac. Example error:
+		// err: <*errors.errorString>{
+		//     s: "no image found in manifest list for architecture \"arm64\", variant \"v8\", OS \"darwin\"",
+		// }
+		sysCtx := &types.SystemContext{
+			OSChoice: "linux",
+		}
```

Comment:
For follow-up:

It looks like what's actually happening right now is that `containers/image` is resolving and pulling an image index based on the tag we provide, and then automatically choosing a manifest from that index.

So that means that by the time we see a descriptor for the image, we are seeing just one manifest, not the index and _all_ of the manifests.

If we could see all of the manifests, then we'd be able to add more checks:
1. Is the root thing actually an index?
2. Does the index have manifests for the 4 architectures we support (amd64, arm64, ppc64le, s390x)?
3. Are the contents of the FBC and cache layers identical for those 4 manifests (could detect just by comparing layer digests)